pkgdb/src/lock-flake-installable.cc

```diff
@@ -0,0 +1,439 @@
+  {
+    auto maybe_outputs_cursor
+      = cursor->findAlongAttrPath( nix::parseAttrPath( *state, "outputs" ) );
+    if ( ! maybe_outputs_cursor )
+      {
+        throw LockFlakeInstallableException(
+          nix::fmt( "could not find '%s.%s' in derivation",
+                    lockedAttrPath,
+                    "outputs" ) );
+      }
+    outputNames = ( *maybe_outputs_cursor )->getListOfStrings();
+
+    for ( auto output : outputNames )
+      {
+        auto outputCursor = cursor->findAlongAttrPath(
+          nix::parseAttrPath( *state, output + ".outPath" ) );
+        if ( ! outputCursor )
+          {
+            throw LockFlakeInstallableException(
+              nix::fmt( "could not find '%s.%s' in derivation",
+                        lockedAttrPath,
+                        output + ".outPath" ) );
+          }
+        auto outputValue = ( *outputCursor )->getStringWithContext();
+        outputs[output]  = outputValue.first;
+      }
+  }
+
+  // determine outputs to install in the following order:
+  // 1. extendedOutputsSpec (`<installable>^out,man`, `<installable>^*`))
+  // 2. `meta.outputsToInstall` field
+  // 3. first output in the `outputs` field
+  std::set<std::string> outputsToInstall;
+  {
+    auto outputSpec = std::visit(
+      overloaded {
+        [&]( const nix::ExtendedOutputsSpec::Default & ) -> nix::OutputsSpec
+        {
+          std::set<std::string> outputsToInstall;
+          auto metaOutputsToInstallCursor = cursor->findAlongAttrPath(
+            nix::parseAttrPath( *state, "meta.outputsToInstall" ) );
+          if ( metaOutputsToInstallCursor )
+            {
+              for ( auto output :
+                    ( *metaOutputsToInstallCursor )->getListOfStrings() )
+                {
+                  outputsToInstall.insert( output );
+                }
+            }
+          else if ( ! outputNames.empty() )
+            {
+              outputsToInstall.insert( outputNames.front() );
+            }
+
+          // this seems to be the default in a few nix places
+          // could also dropt this, since reaching this point means
+          // that the package has no outputs?!
+          if ( outputsToInstall.empty() ) { outputsToInstall.insert( "out" ); }
+          return nix::OutputsSpec::Names { std::move( outputsToInstall ) };
+        },
+        [&]( const nix::ExtendedOutputsSpec::Explicit & e ) -> nix::OutputsSpec
+        { return e; },
+      },
+      extendedOutputsSpec.raw() );
+
+    outputsToInstall = std::visit(
+      overloaded {
+        [&]( const nix::OutputsSpec::Names & n ) -> std::set<std::string>
+        { return n; },
+        [&]( const nix::OutputsSpec::All & ) -> std::set<std::string>
+        {
+          std::set<std::string> outputNamesSet;
+          for ( auto output : outputNames ) { outputNamesSet.insert( output ); }
+          return outputNamesSet;
+        } },
+      outputSpec.raw() );
+  }
+
+  std::string systemAttribute;
```

Comment:
question nonblocking: is this required in derivations? Should we error if it doesn't match the system we're locking for?

Reply:
system is required yes: https://nix.dev/manual/nix/2.23/language/derivations.html#attr-system

As far as i understand, we do want to support (on not restrict at least) any flakeref, i.e. `github:flox/flox#packages.aarch64-linux.flox` should be possible _even_ if the local system is something else.
I do think we need to support a systems attribute on the descriptor in the manifest tho.
While we do defaulting for `packages` and `legacyPackages`, it scurrently impossible to install terivations from anything else without specifying a system and making the env impossible to build on other systems.
That shouldnt be blocking here but id like to keep that in mind!

Reply:
Maybe link that?

If someone has:
```
whatever.flake = github:flox/flox#packages.aarch64-linux.flox
whatever.systems = [ aarch64-linux ]
```
Should we error if we evaluate and `packages.aarch64-linux.flox.system = x86_64-linux`?

Reply:
I guess not? My understanding was that flakes are the escape hatch and restricted as littel as possible

Reply:
Yeah I guess that's fair, if this results in a build failure the user will get an error. Maybe worth a comment